docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,227 @@
+implemented by extending the `GlueShader` class in
+[GlueShader.h](../lgc/elfLinker/GlueShader.h).
+
+When generating an unlinked vertex shader, the entry point will have the name
+`_amdgpu_vs_main_fetchless`. The inputs to this shader will be all of the
+regular inputs to a vertex shader plus all of the vertex input attributes in
+registers. The PAL metadata will contain a table called `.vertexInputs` that
+will pass along information from the original llvm-ir that would otherwise be
+lost, but is needed to load the vertex inputs correctly. When linking, a
+prologue is prepended to the unlinked vertex shader that uses the pipeline state
+and the `.vertexInputs` metadata to load all of the vertex input attributes into
+registers. See the `FetchShader` class in
+[GlueShader.cpp](../lgc/elfLinker/GlueShader.cpp) for the generation of the
+prologue. To see how the unlinked shader handles vertex inputs see the
+`LowerVertexFetch` pass in [VertexFetch.cpp](../lgc/patch/VertexFetch.cpp).
+
+A fragment shader will require a epilogue to export its outputs. The unlinked
+fragment shader will have the standard name for the fragment shader. Instead of
+exporting the outputs, they are left in registers at the end of the shader,
+where there is no `s_endpgm` instruction. Extra information is added to the PAL
+metadata in the `.colorExport` table. During linking this table and the pipeline
+state are used to generate all of the color export instructions and they are
+appended to the end of the unlinked shader. See the `ColorExportShader` class in
+[GlueShader.cpp](../lgc/elfLinker/GlueShader.cpp) for the generation of the
+epilogue. To see how the unlinked shader handles color export see the
+`LowerFragColorExport` pass in
+[FragColorExport.cpp](../lgc/patch/FragColorExport.cpp).
+
+These epilogues and prologues mean that unlinked shaders do not have to depend
+on the format of the vertex input attributes and color exports.
+
+## User data nodes
+
+Consider the following example for the user data nodes:
+![User data node example](user_data_nodes.png)
+
+This is the way the user data nodes are presented to the compiler. Across the
+top are the three "top level" nodes. They are considered the top because the can
+be accessed directly. On the other hand, to access the descriptor for say
+descriptor set 0 and binding 2, the DescriptorTableVsPtr must be loaded first.
+
+Suppose the compiler is generating code to access descriptor set 1 at binding 0
+in a standard compile, where the user data nodes are available. Because it is a
+top level node, the compiler will try to find an input register to hold the
+descriptor. Say it finds `s3`, and it is compiling the vertex shader. Then in
+the PAL metadata the entry `SPI_SHADER_USER_DATA_VS_3` will be assigned the
+offset of the descriptor, which is `0` in this case. Then that descriptor will
+be available in `s3`. Note that the only resource descriptor that can be in the
+top level are buffer descriptors.
```

Comment:
Add a comment that SPI_SHADER_USER_DATA_VS_3 corresponds to s11 on a gfx9+ merged shader? Because merged shaders put system stuff in s0-s7, and user data starts at s8. If we're only doing VS-FS, that doesn't affect us, except that a gfx10 NGG standalone VS is implemented as a GS merged shader.